src/unix/io_uring.c

```diff
@@ -0,0 +1,1340 @@
+    params->sq_off.flags = offsetof(io_rings, sq_flags);
+    params->sq_off.dropped = offsetof(io_rings, sq_dropped);
+    params->sq_off.array = (u8 *)iour->sq_array - (u8 *)iour->rings;
+    runtime_memset((u8 *)params->sq_off.resv, 0, sizeof(params->sq_off.resv));
+    params->cq_off.head = offsetof(io_rings, cq_head);
+    params->cq_off.tail = offsetof(io_rings, cq_tail);
+    params->cq_off.ring_mask = offsetof(io_rings, cq_mask);
+    params->cq_off.ring_entries = offsetof(io_rings, cq_entries);
+    params->cq_off.overflow = offsetof(io_rings, cq_overflow);
+    params->cq_off.cqes = (u8 *)iour->cqes - (u8 *)iour->rings;
+    runtime_memset((u8 *)params->cq_off.resv, 0, sizeof(params->cq_off.resv));
+    return ret;
+err3:
+    deallocate(iour->vh, iour->user_rings, alloc_size);
+err2:
+    deallocate(h, iour->rings, alloc_size);
+err1:
+    deallocate(h, iour, sizeof(*iour));
+    return ret;
+}
+
+sysreturn io_uring_mmap(fdesc desc, u64 len, u64 mapflags, u64 offset)
+{
+    iour_debug("len %ld, flags 0x%x, offset 0x%x", len, mapflags, offset);
+    io_uring iour = (io_uring)desc;
+    u64 region_offset;
+    switch (offset) {
+    case IORING_OFF_SQ_RING:
+    case IORING_OFF_CQ_RING:
+        if (len > IOUR_REGION1_SIZE(iour))
+            return -EINVAL;
+        region_offset = 0;
+        break;
+    case IORING_OFF_SQES:
+        if (len > IOUR_REGION2_SIZE(iour))
+            return -EINVAL;
+        region_offset = IOUR_REGION1_SIZE(iour);
+        break;
+    default:
+        return -EINVAL;
+    }
+    u64 virt = u64_from_pointer(iour->user_rings) + region_offset;
+    map(virt, physical_from_virtual(iour->rings) + region_offset, len,
+        mapflags);
+    return virt;
+}
+
+simple_closure_function(1, 2, void, iour_efd_complete,
+                        u64, efd_val,
+                        thread, t, sysreturn, rv)
+{
+    closure_finish();
+}
+
+static void iour_complete_locked(io_uring iour, u64 user_data, s32 res,
+                                 boolean async)
+{
+    io_rings rings = iour->rings;
+    iour_debug("user_data %ld, res %d, CQ tail %d", user_data, res,
+               rings->cq_tail);
+    if (rings->cq_tail < rings->cq_head + iour->cq_entries) {
+        struct io_uring_cqe *cqe = &iour->cqes[rings->cq_tail & iour->cq_mask];
+        cqe->user_data = user_data;
+        cqe->res = res;
+        cqe->flags = 0;
+        write_barrier();
+        rings->cq_tail++;
+    } else {
+        iour_debug("overflow");
+        rings->cq_overflow++;
+    }
+    if (iour->eventfd && (async || !iour->eventfd_async)) {
+        closure_new(iour->h, iour_efd_complete, completion);
+        if (completion != INVALID_ADDRESS) {
+            completion->efd_val = 1;
+            apply(iour->eventfd->write, &completion->efd_val,
+                  sizeof(completion->efd_val), 0, current, true,
+                  closure_get(iour_efd_complete, completion));
+        }
```

Comment:
assert or otherwise handle alloc error here?

Reply:
The allocation is actually done by `closure_new()` (after which I check for `completion != INVALID_ADDRESS`, i.e. check against allocation failure). `closure_get()` is just synctatic sugar to retrieve the apply() function from the closure name.